### PR TITLE
[Feature] Add `update_repo_settings` function to HfApi #2447

### PR DESCRIPTION
**Pull request by WizKnight**
This PR adds a new function `update_repo_settings` to the `HfApi` class, allowing users to programmatically update various repository settings, including gating.

### CODE REVIEW

**Comment by WizKnight** [comment]
Hey @Wauplin, can you please go through this added changes and let me know if I'm in the right direction??
Also the tests are failing locally with this changes.

File: src/huggingface_hub/hf_api.py
Comment (hanouticelina) [inline src/huggingface_hub/hf_api.py:3545]: I think this method should be renamed to something like `update_repo_gated_status`. imo, It better describes the function's purpose of modifying a repo's gated access status.

**Review by hanouticelina — CHANGES_REQUESTED** [review]
Hello @WizKnight, thank you for this PR :) I left few comments, also I noticed that `HfApi.upload_large_folder` has been removed. This function is still present in the main branch and appears to be in use. Did you remove it for some reason?

**Comment by WizKnight** [comment]
> Hello @WizKnight, thank you for this PR :) I left few comments, also I noticed that `HfApi.upload_large_folder` has been removed. This function is still present in the main branch and appears to be in use. Did you remove it for some reason?

After updating my local branch, I noticed that `HfApi.upload_large_folder` was automatically removed. Upon cross-checking with the main branch, I found that a new file named [_upload_large_folder.py](https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/_upload_large_folder.py) had been created.

File: src/huggingface_hub/hf_api.py
Comment (WizKnight) [reply src/huggingface_hub/hf_api.py:3545]: Yeah, `update_repo_gated_status` this will have more clarity

Comment (Wauplin) [reply src/huggingface_hub/hf_api.py:3545]: Hi, sorry the naming comes from me in https://github.com/huggingface/huggingface_hub/issues/2447#issue-2465284588 and I haven't provided the context for it. Repos have multiple settings (gating, visibility, discussion tab, etc.) and I think it'll be nice to have a single helper to handle them all. Defining the method `update_repo_settings` in the PR to handle gating is the first step towards this. I think it'll make sense in a future PR to deprecate `update_repo_visibibility` and/or make it an alias of this new one.

Comment (hanouticelina) [reply src/huggingface_hub/hf_api.py:3545]: Makes sense! 👍

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:3549]:
```suggestion
        gated: Literal["auto", "manual", False] = False,
```

(even more precise like this since `True` is not handled)

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:3555]:
```suggestion
            raise ValueError(f"Invalid gated status, must be one of 'auto', 'manual', or False. Got '{gated}'".)
```

(nit)

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:3573]: `hf_raise_for_status` already handles error codes like 401 or 404 with correct messages and exceptions so no need to re-implement it. In this case, you can simply do

```py
        hf_raise_for_status(r)
```
and that's enough. In general, it's good practice to raise the root error to the user (here a HTTPError), rather than a `ValueError`.

File: tests/test_hf_api.py
Comment (Wauplin) [inline tests/test_hf_api.py:230]: Instead of testing the return value (which is best to remove), you can do 


```py
res = self._api.model_info(repo_id, expand="gated")
```

This will retrieve the information from the server, hence confirming it has been correctly taken into account.

File: tests/test_hf_api.py
Comment (Wauplin) [inline tests/test_hf_api.py:304]: To be removed once the PR is fully ready

File: tests/test_hf_api.py
Comment (Wauplin) [inline tests/test_hf_api.py:142]: No need to modify this test (it's not related to your feature)

**Review by Wauplin — COMMENTED** [review]
Hey there, thanks for your contribution @WizKnight! I've pushed some commits to fix your merging issues. I don't know what happened but it looks like you did not start a new branch from the `main` branch (but maybe from your previous one?) and then a merge conflict did not go well. Anyway, should be good now if you pull the latest changes.

Also posted a few more comments in addition to @hanouticelina 's review. Let us know if you have any question :)

File: src/huggingface_hub/hf_api.py
Comment (WizKnight) [reply src/huggingface_hub/hf_api.py:3575]: The `else` block here handles any `HTTPError` raised by the `hf_raise_for_status(r)` call that doesn't fall into the specific cases of 401 (Unauthorized) or 404 (Not Found) errors

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:3550]:
```suggestion
        *,
        gated: Literal["auto", "manual", False] = False,
```

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py:3571]: For now we can just avoid having a return value at all.

```suggestion
```

File: src/huggingface_hub/hf_api.py
Comment (WizKnight) [reply src/huggingface_hub/hf_api.py:3550]: Okay, only these two modifications are required?

Comment (Wauplin) [reply src/huggingface_hub/hf_api.py:3550]: I think so but I'll have to re-review more in depth once applied (I did not try it locally yet)

Comment (WizKnight) [reply src/huggingface_hub/hf_api.py:3550]: Okay, I'll commit these changes. After you've tested them locally, feel free to let me know if any further adjustments are needed.

File: docs/source/en/guides/repository.md
Comment (Wauplin) [inline docs/source/en/guides/repository.md:158]:
```suggestion
### Setup gated access
```

(this section is already under `## Change repository settings`)

File: docs/source/en/guides/repository.md
Comment (Wauplin) [inline docs/source/en/guides/repository.md:161]:
```suggestion
To give more control over how repos are used, the Hub allows repo authors to enable **access requests** for their repos. Users must agree to share their contact information (username and email address) with the repo authors to access the  files when enabled. A repo with access requests enabled is called a **gated repo**.

You can set a repo as gated using [`update_repo_settings`]:
```

I took over the existing wording from https://huggingface.co/docs/hub/models-gated introduction to be more consistent in our wording.

File: docs/source/en/guides/repository.md
Comment (Wauplin) [inline docs/source/en/guides/repository.md:164]:
```suggestion
```

Except very rare exceptions, methods of `HfApi` must be primarily designed to work on model repos.

File: docs/source/en/guides/repository.md
Comment (Wauplin) [inline docs/source/en/guides/repository.md:169]:
```suggestion
>>> api.update_repo_settings(repo_id=repo_id, gated="auto")  # Set automatic gating for a model
```

File: tests/test_hf_api.py
Comment (Wauplin) [inline tests/test_hf_api.py:305]:
```suggestion
    @use_tmp_repo(repo_type="model")
    def test_update_repo_settings(self, repo_url: RepoUrl):
        repo_id = repo_url.repo_id

        for gated_value in ["auto", "manual", False]:
            self._api.update_repo_settings(repo_id=repo_id, gated=gated_value)
            info = self._api.model_info(repo_id, expand="gated")
            assert info.gated == gated_value
```

The logic of the test can be made a bit more concise by:
- using `@use_tmp_repo` that will handle repo creation and deletion for you. With the current implementation if an error occurs, `self._api.delete_repo` is not called to clean up the test repo. The decorator handle both if the test succeeds or fails
- we can make the logic to test all 3 values a bit more compact given very little difference between each case. Writing tests is always a balance between simplicity and readability. In the case here, it's fine to add a for loop if it saves 6 duplicated lines of code

File: tests/test_hf_api.py
Comment (Wauplin) [inline tests/test_hf_api.py:337]: This is already tested above

File: tests/test_hf_api.py
Comment (Wauplin) [inline tests/test_hf_api.py:343]: No need to test this here. Usually tests should test only 1 thing. Repo creation is being tested elsewhere so here we want to focus only on `update_repo_settings`.

**Review by Wauplin — COMMENTED** [review]
Hi @WizKnight, thanks for the iteration! I've re-reviewed your work and left a few more comments. Once those are addressed, we should be good to merge :)

File: tests/test_hf_api.py
Comment (WizKnight) [reply tests/test_hf_api.py:305]: That's a great suggestion! I appreciate your feedback on how to improve test logic.

Using the `@use_tmp_repo` decorator definitely makes the test code more concise and ensures proper cleanup even in case of errors.
Regarding the second point, I agree that we can compact the logic to test all three values of gated using a for loop. This would eliminate code duplication and enhance readability.

I'll keep these suggestions in mind when writing future tests. Thanks again for your input!

**Review by Wauplin — COMMENTED** [review]
Thanks for the changes @WizKnight! Everything looks good to me now logic-wise :) 

It looks like you still have some minor styling issues. Can you pull the latest version (I've merged from `main`), run `make style` locally to format the code, run `make quality` to check everything's correct and then commit the changes. This should solve the current issues. Thanks!

**Comment by WizKnight** [comment]
> Thanks for the changes @WizKnight! Everything looks good to me now logic-wise :)
> 
> It looks like you still have some minor styling issues. Can you pull the latest version (I've merged from `main`), run `make style` locally to format the code, run `make quality` to check everything's correct and then commit the changes. This should solve the current issues. Thanks!

After the `make quality` WSL is giving error and asking to put in return value for the function. So should I use `return r.json()` ??

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:3585]:
```suggestion
    ) -> None:
```


@WizKnight I think your issue comes from the fact that the method expects a return value here. Since we're not returning anything, better to annotate with ` -> None`

**Review by hanouticelina — CHANGES_REQUESTED** [review]
Hi @WizKnight, thanks so much for your interations on this PR! 🤗 I've left some comments regarding the failing tests

File: tests/test_hf_api.py
Comment (Wauplin) [reply tests/test_hf_api.py:313]: I think we can drop the test for Space repo_type then, thanks for looking into it @hanouticelina :) Testing on model and dataset is already more than enough :+1:

Comment (WizKnight) [reply tests/test_hf_api.py:313]: Okay then @Wauplin :) , I'll remove the space_repo_type test and commit again.

Comment (WizKnight) [reply tests/test_hf_api.py:306]: Hey @hanouticelina, thanks for taking a look at the `space_repo_settings` test. I removed it in the latest commit, as @Wauplin recommended. Also, I added `repo_type` to `dataset_repo_settings` based on your suggestion.

**Review by Wauplin — COMMENTED** [review]
Thanks @WizKnight! Almost looking good now :) I just realized there is a docstring missing. Could you add one to the method so that it's well documented for users? You can take inspiration from [this one](https://github.com/huggingface/huggingface_hub/blob/26c2d89f5e521b50916fbaf190e5a42705bcb775/src/huggingface_hub/hf_api.py#L3684). No need to write one as complete as this one but at least have:
1. a short explanation
2. an `args` section
3. a `raises` section

I promise, this is the last step to have a clean and self-contained PR!

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:3581]:
```suggestion
            repo_id (`str`):
```

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:3583]:
```suggestion
            gated (`Literal["auto", "manual", False]`, *optional*):
```

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:3587]:
```suggestion
                * False (default): The repository is not gated, and anyone can access it.
```

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:3588]:
```suggestion
            token (`Union[str, bool, None]`, *optional*):
```

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:3593]:
```suggestion
            repo_type (`str`, *optional*):
```

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:3597]:
```suggestion
                The type of the repository to update settings from (`"model"`, `"dataset"` or `"space"`.
                Defaults to `"model"`.
```

**Review by Wauplin — COMMENTED** [review]
minor comments

**Review by Wauplin — APPROVED** [review]
Everything looks good to me! Thanks for your patience @WizKnight! :smile:
Let's wait for a last review from @hanouticelina and we should be good :)

**Review by hanouticelina — APPROVED** [review]
Looks good to me too! thanks again @WizKnight for this! 🎉  The failing test is not related to your PR, let me check that

**Comment by WizKnight** [comment]
Thank you so much, @Wauplin & @hanouticelina !! 🤗 I truly appreciate your guidance and patience throughout this process.

@Wauplin, I'm constantly learning from you, and I'm excited to tackle new challenges. I'd be grateful if you could assign me another task once this PR is successfully closed.

**Comment by hanouticelina** [comment]
Failing tests are unrelated so let's get this merged :)

**Comment by Wauplin** [comment]
Hi @WizKnight, I just created https://github.com/huggingface/huggingface_hub/issues/2537 which would be the follow-up work to do after this PR. Interested in working on it?

**Comment by WizKnight** [comment]
> Hi @WizKnight, I just created #2537 which would be the follow-up work to do after this PR. Interested in working on it?

Hi @Wauplin, I would love to. This looks interesting!!

**Pull request merged**
